lib/wasi/src/runtime/module_cache/types.rs

```diff
@@ -0,0 +1,141 @@
+use std::{
+    fmt::{self, Debug, Display, Formatter},
+    ops::Deref,
+};
+
+use sha2::{Digest, Sha256};
+use wasmer::{Engine, Module};
+
+use crate::runtime::module_cache::AndThen;
+
+/// A cache for compiled WebAssembly modules.
+///
+/// ## Assumptions
+///
+/// Implementations can assume that cache keys are unique and that using the
+/// same key to load or save will always result in the "same" module.
+///
+/// Implementations can also assume that [`ModuleCache::load()`] will
+/// be called more often than [`ModuleCache::save()`] and optimise
+/// their caching strategy accordingly.
+#[async_trait::async_trait]
+pub trait ModuleCache: Debug {
+    async fn load(&self, key: Key, engine: &Engine) -> Result<Module, CacheError>;
+
+    async fn save(&self, key: Key, module: &Module) -> Result<(), CacheError>;
+
```

Comment:
Should probably have a `remove(hash)` and a `purge()/clear()`.

Reply:
I'm tempted to say it's not the `ModuleCache` trait's responsibility to remove items from a cache or purge them because those extra methods won't be used by WASIX. 

There's nothing stopping us from creating a `DeployModuleCache` and adding `remove()`, `purge()`, and `clear()` as normal methods, though.

Reply:
I kind of like the idea of having `delete(key, &engine)` and `purge(&engine)`

Reply:
I agree it could be useful, but I don't want to start adding extra complexity to the `ModuleCache` trait until we see a need for it. 

Also keep in mind that the `ModuleCache` trait doesn't need to include everything any cache implementation could ever need. It's perfectly fine for cache implementations to have their own inherent methods which do operations that make sense for that cache. It's only when you need to do things in a generic context that they need to be added to the trait.